chapters/part3/document-ranking.md

```diff
@@ -0,0 +1,350 @@
+
+The second is how to handle labels. While labels have normally been discrete, we now have continuous
+scores. For this reason, we'll forego the use of a `LabelField`. The main utility of a `LabelField`
+is to index our labels for easy conversion, but we don't have string labels here that we need
+indexed. We also have multiple labels. If we wanted to, we could use a `ListField[LabelField]` here,
+but instead, we're just going to use an
+[`ArrayField`](https://docs.allennlp.org/master/api/data/fields/array_field/):
+
+> A class representing an array, which could have arbitrary dimensions. A batch of these arrays are
+> padded to the max dimension length in the batch for each dimension.
+
+This is a better fit for our dataset. Here's an overview of our `text_to_instance` method; for the
+full dataset reader, see the code
+[here](https://github.com/jacobdanovitch/allenrank/blob/master/allenrank/dataset_readers/mimics_reader.py).
+
+```python
+
+    @overrides
+    def text_to_instance(
+        self,
+        query: str,
+        question: str,
+        options: List[str],
+        labels: List[float] = None
+    ) -> Instance:
+
+        token_field = self._make_textfield((query, question))
+        options_field = self._make_listfield(options)
+
+        fields = { 'tokens': token_field, 'options': options_field }
+        if labels:
+            fields['labels'] = ArrayField(np.array(labels), padding_value=-1)
+
+        return Instance(fields)
+```
+
+
+</exercise>
+
+
+<exercise id="3" title="Designing our ranking model">
+
+Let's start off by understanding our inputs and outputs. Our model is going to accept `tokens` (the
+concatenation of the query and question), `options` (the `ListField[TextField]` of answer
+candidates), and `labels` (the `ArrayField` containing a continuous score for each option). The
+sizes have been annotated inline.
+
+```python
+@Model.register("ranker")
+class DocumentRanker(Model):
+
+    # ...
+
+    def forward(
+        self,
+        tokens: Dict[str, torch,Tensor], # batch * words
+        options: List[Dict[str, torch,Tensor]], # batch * num_options * words
+        labels: torch.FloatTensor = None # batch * num_options
+    ) -> Dict[str, torch.Tensor]:
+```
+
+Now, how do we convert our token IDs to embeddings when they're in `ListField` form?
+`TextFieldEmbedders` facilitate this by accepting an extra `num_wrapping_dims` keyword argument.
+
+```python
+    def forward(
+        self,
+        tokens: Dict[str, torch,Tensor], # batch * words
+        options: List[Dict[str, torch,Tensor]], # batch * num_options * words
+        labels: torch.FloatTensor = None # batch * num_options
+    ) -> Dict[str, torch.Tensor]:
+        embedded_text = self._text_field_embedder(tokens)
+        mask = get_text_field_mask(tokens).long()
+
+        embedded_options = self._text_field_embedder(options, num_wrapping_dims=1)
+        options_mask = get_text_field_mask(options).long()
```

Comment:
Should I include a comment noting that 0 is the default, so it's not strictly necessary?

Reply:
This shouldn't be working without `num_wrapping_dims=1`.  If it is, I'm really not sure why.  Maybe because the transformer is creating a `mask` tensor that is already correctly wrapped?